### ISSUE TITLE
RFC: Product Service Duration

### ISSUE THREAD

**Issue by lasley** [OCA/product-attribute#206]
This module will add the concept of a service user_id and duration on `product.template`. 

The intent is to allow for the creation of services that are (optionally) performed by a certain user & have a standard duration in minutes.

The specific need on my end is to use as part of Physician scheduling functionality, pairing with the calendar module to allow selection of the service that is going to be performed (office visit, etc) in order to default a time block size and optional user. 

This seems useful in a lot of other cases though; basically any form of job scheduling where X units takes Y minutes of time.

Assuming I'm not missing some strategy that already exists for this, am I also in the right repo?

**Comment by pedrobaeza** [comment]
For this, you already have as a base https://github.com/OCA/sale-workflow/tree/10.0/sale_start_end_dates, and sale_rental has a similar concept, but it doesn't include yet the scheduling.

**Comment by lasley** [comment]
Thanks for the suggestion @pedrobaeza - I thought about `sale_start_end_dates` at first, but most of this logic is incompatible with my needs unfortunately. It is focused around dates on sales and invoices, while I need just a float_time on products dictating how long that service generally takes to perform (and possibly a person that performs it).

An example is a Doctor's Office visit, which will typically takes say 30 minutes. The scheduling assistant can select the associated product for an office visit during scheduling, which will enforce a 30 minutes time block for the appointment. The product can then be used for subsequent billing when the appointment is actually handled.

Another example in a different context with no need for billing the appointment, would be a restaurant taking reservations. They know that the average reservation for a party is 2 hours, so they use this service product to reserve the time block in the calendar. Maybe there is a cancellation fee associated with this product that is used for billing, but for the most part the billing is on the food itself.

In both examples these workflows are taking place from a calendar, and don't need the creation of a sale order during scheduling. The amount of time taken for the appointment isn't relevant to anyone past the scheduling, so doesn't really flow into the billing processes like the existing modules.

As I typed this out, I'm thinking maybe `event` is a better repo. While I did not originally intend for this base module to depend on event, or provide that logic, I'm having troubles thinking of what else it would be good for.

**Comment by pedrobaeza** [comment]
I think you should depend on `resource` module and base all your work to think in a load/capacity planner. We have this in our R&D projects, but to be plugged to projects and manufacturing, but the problem is the same: you have several resources (a table, a doctor, an engineer, a workcenter...), and you need it for a time. If it's free, you can lock it for a time slot.

So first of all I would develop a generic toolset around resource model for this (you already have a pretty decent toolset for it on the module itself).

Then, you need to convert your objects into resources (inheriting from this model), and then the only that rests is some generic wizards/views for planning, calculating agendas, and so on.

What do you think?

**Comment by lasley** [comment]
We're totally on the same page now 😄 . Thinking about it even more abstract, this would also allow for the utilization & scheduling of non-human resources without work orders.

Looking at resource, it seems a big missing piece is a link between a resource calendar and an event calendar. Do you know of anything that glues these two modules together?

**Comment by pedrobaeza** [comment]
The only thing that exists right now is https://github.com/OCA/crm/pull/118.

**Comment by lasley** [comment]
Thanks @pedrobaeza - much appreciated, as always. I'll keep this RFC open in case anyone else has comments & will likely have a PR submitted here in the next few weeks.

**Comment by pedrobaeza** [comment]
Odoo is also planning a similar functionality: https://youtu.be/XyHliL_t9xU

**Comment by lasley** [comment]
Oh neat, that functionality looks similar to Doodle.com. Not quite what I'm going for here though, because a scheduling assistant would be making the appointment immediately against a known calendar vs. sending it to participants to select the time. Any idea where this functionality is (or will be)?

**Comment by pedrobaeza** [comment]
@lasley, any news/PR on this?

**Comment by lasley** [comment]
We have an internal ticket for dev, but we were waiting on OCA/crm#118 to be finished so that we could use some of the logic there. I went ahead and made a ticket to finish calendar_resource, seems like that's the best course of action.

OCA/crm#118 ticket is now scheduled in our sprint that starts next week, and product_service_duration the following sprint.

**Comment by pedrobaeza** [comment]
OK, I keep this open then.

**Comment by woodbrettm** [comment]
OCA/crm#118 is now superseded by OCA/crm#135

**Comment by lasley** [comment]
@bmw95 We'll wait to see if @dreispt wants to pick up the changes in his v8 PR before we consider it truly superseded.

There's still another portion of this RFC that hasn't been accomplished yet, so I am still going to leave open.

**Comment by lasley** [comment]
Closing to track in #251
